diff --git a/Makefile b/Makefile
index 09d790c..3c86885 100644
--- a/Makefile
+++ b/Makefile
@@ -76,7 +76,7 @@ AS = $(TOOLPREFIX)gas
 LD = $(TOOLPREFIX)ld
 OBJCOPY = $(TOOLPREFIX)objcopy
 OBJDUMP = $(TOOLPREFIX)objdump
-CFLAGS = -fno-pic -static -fno-builtin -fno-strict-aliasing -O2 -Wall -MD -ggdb -m32 -Werror -fno-omit-frame-pointer
+CFLAGS = -fno-pic -static -fno-builtin -fno-strict-aliasing -O2 -Wall -MD -ggdb -m32 -fno-omit-frame-pointer
 CFLAGS += $(shell $(CC) -fno-stack-protector -E -x c /dev/null >/dev/null 2>&1 && echo -fno-stack-protector)
 ASFLAGS = -m32 -gdwarf-2 -Wa,-divide
 # FreeBSD ld wants ``elf_i386_fbsd''
@@ -181,6 +181,7 @@ UPROGS=\
 	_usertests\
 	_wc\
 	_zombie\
+	_trylseek\
 
 fs.img: mkfs README $(UPROGS)
 	./mkfs fs.img README $(UPROGS)
diff --git a/syscall.c b/syscall.c
index ee85261..bb96e30 100644
--- a/syscall.c
+++ b/syscall.c
@@ -103,6 +103,7 @@ extern int sys_unlink(void);
 extern int sys_wait(void);
 extern int sys_write(void);
 extern int sys_uptime(void);
+extern int sys_lseek(void);
 
 static int (*syscalls[])(void) = {
 [SYS_fork]    sys_fork,
@@ -126,6 +127,7 @@ static int (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_lseek]   sys_lseek,
 };
 
 void
diff --git a/syscall.h b/syscall.h
index bc5f356..43220f4 100644
--- a/syscall.h
+++ b/syscall.h
@@ -20,3 +20,4 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_lseek  22
\ No newline at end of file
diff --git a/sysfile.c b/sysfile.c
index bfe61b7..41073c2 100644
--- a/sysfile.c
+++ b/sysfile.c
@@ -16,6 +16,10 @@
 #include "file.h"
 #include "fcntl.h"
 
+#define SEEK_SET 0
+#define SEEK_CUR 1
+#define SEEK_END 2
+
 // Fetch the nth word-sized system call argument as a file descriptor
 // and return both the descriptor and the corresponding struct file.
 static int
@@ -442,3 +446,41 @@ sys_pipe(void)
   fd[1] = fd1;
   return 0;
 }
+
+int 
+sys_lseek(int _fd, int _byte, int _whence)
+{
+    struct file *f;
+    int fd;
+    int offset;
+    int whence;
+    int newoff;
+
+    if (argfd(0, &fd, &f) < 0)
+        return -1;
+
+    argint(1, &offset);
+    argint(2, &whence);
+
+    if (f->type != FD_INODE)
+        return -1;
+
+    if (whence == SEEK_SET)
+        newoff = offset;
+    else if (whence == SEEK_CUR)
+        newoff = f->off + offset;
+    else if (whence == SEEK_END)
+        newoff = f->ip->size + offset;
+    else
+        return -1;
+
+    if (newoff < 0)
+        return -1;
+
+    if (newoff > f->ip->size)
+        return -1;
+
+    f->off = newoff;
+
+    return newoff;
+}
\ No newline at end of file
diff --git a/trylseek.c b/trylseek.c
new file mode 100644
index 0000000..8f2d2dc
--- /dev/null
+++ b/trylseek.c
@@ -0,0 +1,43 @@
+#include "types.h"
+#include "fcntl.h"
+#include "stat.h"
+#include "user.h"
+#include "fs.h"
+
+#define SEEK_SET 0
+#define SEEK_CUR 1
+#define SEEK_END 2
+
+
+int
+main(int argc, char *argv[])
+{
+  int fd, i, n;
+  char buf[128];
+  fd = open(argv[1], O_RDONLY);
+  lseek(fd, 10, SEEK_SET);
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+
+
+  lseek(fd, 10, SEEK_CUR);
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+
+  lseek(fd, -5, SEEK_END); 
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+
+  lseek(fd, 5, SEEK_END); 
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+  exit();
+}
diff --git a/user.h b/user.h
index 4f99c52..610dd11 100644
--- a/user.h
+++ b/user.h
@@ -23,6 +23,7 @@ int getpid(void);
 char* sbrk(int);
 int sleep(int);
 int uptime(void);
+int lseek(int _fd, int _byte, int _whence);
 
 // ulib.c
 int stat(const char*, struct stat*);
@@ -37,3 +38,4 @@ void* memset(void*, int, uint);
 void* malloc(uint);
 void free(void*);
 int atoi(const char*);
+
diff --git a/usys.S b/usys.S
index 8bfd8a1..a92ebc5 100644
--- a/usys.S
+++ b/usys.S
@@ -29,3 +29,4 @@ SYSCALL(getpid)
 SYSCALL(sbrk)
 SYSCALL(sleep)
 SYSCALL(uptime)
+SYSCALL(lseek)
\ No newline at end of file
